Several fixes to the parser and syntax feedback

guillep opened the pull request:
This PR introduces several enhancements, fixes and cleanups.
 - refactoring big methods into small ones (specially taking out code from loops to make debugging easier)
 - fixing the parser so it correctly consumes all the stream of tokens
 - fixing the parser so each scope (methods, blocks, curly braces) stop consuming statements as soon as they find the right token

 - extended the faulty parser mode to recognize unfinished statements (e.g., missing dots)
 - extended the faulty parser mode to recognize unreachable statements (i.e., statements following a return)

On the icon style side:
 - englobing errors (unfinished statements, missing closers/openers of parenthesis) only underline the last character
   => less noise
 - unreachable statements are completely underlined
 - fixed the SHOUT syntax highlighter to highlight inside error nodes that wrap normal nodes

A taste of it:
![image](https://user-images.githubusercontent.com/708322/94971459-5c840d00-0507-11eb-8353-91d4ac4af110.png)

Compare to Pharo 8:

![image](https://user-images.githubusercontent.com/708322/94987707-90434f00-0568-11eb-9751-22a686240323.png)

And to Pharo9 previous to this change:

![image](https://user-images.githubusercontent.com/708322/94987710-9afde400-0568-11eb-8316-882080429f59.png)

guillep committed:
Introduce the concept of Unreachable statement and unfinished statement

guillep committed:
Add convenience methods to instantiate the parser

guillep committed:
Use the value of the token and to avoid using the source

guillep committed:
Extract parseStatement from parseStatement list to make it more readable

guillep committed:
Correct statement parsing stop conditions depending on the current scope e.g., blocks should finish on $]

guillep committed:
Englobing error node are parse error nodes too

guillep committed:
Do not override the selector of methods

guillep committed:
Fix tests with new error parsing to take into account unfinished statements and unreachable statements

guillep committed:
Cleanup: extract parsing of temporaries to make debugging easier

guillep committed:
Refactor to remove duplicated parseStatements:into:

guillep committed:
Extract parse cascade message to simplify debugging, and extend parse error nodes to be compatible with message nodes

guillep committed:
Syntax highlighter should highlight inside englobing errors

guillep committed:
Special case for styling englobing error nodes: only highlight the last character of it

guillep committed:
Extending the ErrorNodeStyler to underline correctly unreachable statements

guillep committed:
Better cascade error message

guillep committed:
A faulty cascade message should only consume a token if it is not a closer token or a cascade separator.
Add a comment, and enhance the error message.

guillep committed:
Fix compiler expected messages (the new ones are more intention revealing)
Fix assertion to put expectation at the right

Ducasse commented:
Guille these are failing. 

- testParseError2 – Unix64.EnlumineurFormatter.Tests.EFParseErrorExpressionTest

Ducasse commented:
- testBestNodeWithInvalidPlaygroundUnParsableError – Unix64.Rubric.Tests.Base.RubSmalltalkEditorTest

guillep commented:
Thanks, I'll take a look later, the one of rubric looks scary :)

guillep committed:
Fix failing tests.
 - remove rubric test: it should not duplicate the AST semantics, ASTs are already tested in the AST package
 - patch formatter test: we should revise error nodes since they do not always carry the right information to be reformatted
